### PR TITLE
[mlir][SME] Add vector.splat -> SME conversion

### PR DESCRIPTION
**Pull request by banach-space**
This conversion is identical to vector.broadcast when broadcasting a
scalar.

### CODE REVIEW

File: mlir/lib/Conversion/VectorToArmSME/VectorToArmSME.cpp [hunk 1]
```diff
@@ -240,6 +240,63 @@ struct BroadcastOpToArmSMELowering
...
+
+    auto srcType = splatOp.getOperand().getType();
+    auto tileElementType = tileType.getElementType();
+
+    assert(srcType.isIntOrFloat() && "Invalid source type for vector.splat");
+
+    // First, broadcast the scalar to a 1-d vector.
+    auto tileSliceType =
+        VectorType::get(tileType.getShape().drop_front(), tileElementType,
+                        /*scalableDims=*/{true});
```
Comment (MacDue) [inline mlir/lib/Conversion/VectorToArmSME/VectorToArmSME.cpp:279]:
```suggestion
    VectorType tileSliceType = VectorType::Builder(tileType).dropDim(0);
```

File: mlir/lib/Conversion/VectorToArmSME/VectorToArmSME.cpp [hunk 2]
```diff
@@ -240,6 +240,63 @@ struct BroadcastOpToArmSMELowering
...
+struct SplatOpToArmSMELowering : public OpRewritePattern<vector::SplatOp> {
+  using OpRewritePattern<vector::SplatOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(vector::SplatOp splatOp,
+                                PatternRewriter &rewriter) const final {
+    auto tileType = splatOp.getResult().getType();
+    if (!tileType || !arm_sme::isValidSMETileVectorType(tileType))
+      return failure();
+
+    OpBuilder::InsertionGuard g(rewriter);
```
Comment (MacDue) [inline mlir/lib/Conversion/VectorToArmSME/VectorToArmSME.cpp:268]: Is this guard necessary? It exists for the entire scope of the rewrite

**Review by MacDue — APPROVED** [review]
LGTM, just some nits:

File: mlir/lib/Conversion/VectorToArmSME/VectorToArmSME.cpp [hunk 3]
```diff
@@ -240,6 +240,63 @@ struct BroadcastOpToArmSMELowering
...
+/// is converted to:
+///
+///   %broadcast_to_1d = vector.broadcast %src : i32 to vector<[4]xi32>
+///   scf.for %tile_slice_index = %c0 to %num_tile_slices step %c1 {
+///     arm_sme.move_vector_to_tile_slice %broadcast_to_1d, %tile,
+///       %tile_slice_index : vector<[4]xi32> into vector<[4]x[4]xi32>
+///   }
+///
+/// This should, in practice, be identical to vector.broadcast when
+/// broadcasting a scalar.
```
Comment (c-rhodes) [inline mlir/lib/Conversion/VectorToArmSME/VectorToArmSME.cpp:258]: It's identical, I think the ambiguity can be removed
```suggestion
/// This is identical to vector.broadcast of a scalar.
```

**Review by c-rhodes — COMMENTED** [review]
thanks for patch Andrzej, one more vector op we can tick of the list

**Review by c-rhodes — APPROVED** [review]
LGTM cheers

**Pull request merged**